test/utils/test_jinja_chat_extension.py

```diff
@@ -0,0 +1,339 @@
+        assert output == expected
+
+    def test_invalid_role(self, jinja_env):
+        template = """
+        {% message role="invalid_role" %}
+        This should fail
+        {% endmessage %}
+        """
+        with pytest.raises(TemplateSyntaxError, match="Role must be one of"):
+            jinja_env.from_string(template).render()
+
+    def test_for_template_filter_with_invalid_type(self):
+        with pytest.raises(ValueError, match="Value must be an instance of one of the following types"):
+            for_template(123)
+
+    def test_empty_message_content_raises_error(self, jinja_env):
+        template = """
+        {% message role="user" %}
+        {% endmessage %}
+        """
+        with pytest.raises(ValueError, match="Message content is empty"):
+            jinja_env.from_string(template).render()
+
+
+    def test_message_with_whitespace_handling(self, jinja_env):
+        # the following templates should all be equivalent
+        templates = [
+            """{% message role="user" %}{{ image | for_template }}{% endmessage %}""",
+            """{% message role="user" %}    {{ image | for_template }}    {% endmessage %}""",
+            """{% message role="user" %}
+            {{ image | for_template }}
+            {% endmessage %}""",
+            """{% message role="user" %}\t{{ image | for_template }}\t{% endmessage %}"""
+        ]
+        image = ImageContent(base64_image="test_base64", mime_type="image/jpeg")
+        expected = {
+            "role": "user",
+            "content": [
+                {"image": {
+                    "base64_image": "test_base64",
+                    "mime_type": "image/jpeg",
+                    "detail": None,
+                    "meta": {}
+                }}
+            ],
+            "name": None,
+            "meta": {}
+        }
+        for template in templates:
+            rendered = jinja_env.from_string(template).render(image=image)
+            output = json.loads(rendered.strip())
+            assert output == expected
+
+    def test_unclosed_content_tag_raises_error(self, jinja_env):
+        template = """
+        {% message role="user" %}
+        <haystack_content_part>{"type": "text", "text": "Hello"}
+        {% endmessage %}
+        """
+        with pytest.raises(ValueError, match="Found unclosed <haystack_content_part> tag"):
+            jinja_env.from_string(template).render()
+
+
+    def test_invalid_json_in_content_part_raises_error(self, jinja_env):
+        template = """
+        {% message role="user" %}
+        Normal text before.
+        <haystack_content_part>{"this is": "invalid" json}</haystack_content_part>
+        <haystack_content_part>not even trying to be json</haystack_content_part>
+        <haystack_content_part>{]</haystack_content_part>
+        Normal text after.
+        {% endmessage %}
+        """
+        with pytest.raises(json.JSONDecodeError):
+            jinja_env.from_string(template).render()
+
+
```

Comment:
Could we also add some tests for when a user forgets a `{% endmessage %}` tag? I'd also be curious to know what happens in this case and what the error message would be.

Reply:
`jinja2.exceptions.TemplateSyntaxError: Unexpected end of template. Jinja was looking for the following tags: 'endmessage'. The innermost block that needs to be closed is 'message'.`

Test added in https://github.com/deepset-ai/haystack-experimental/pull/297/commits/e123c9b544dcbb99469f670412eabf7d082827bb